### DISCUSSION TITLE
[OpenCV4] Cannot use vcpkg_check_features() for "dnn", "gtk", ipp", "openmp", "ovis", "python", "qt", "tbb"

### DISCUSSION THREAD

**Discussion by gblikas** [microsoft/vcpkg#40455] [General]
In the [latest opencv4 portfile](https://github.com/microsoft/vcpkg/blob/master/ports/opencv4/portfile.cmake) there is the line https://github.com/microsoft/vcpkg/blob/333ba63a16024f05f9172fe403f9eced000389e5/ports/opencv4/portfile.cmake#L98-L99

A quick search for `\"dnn\"` on github, via `/` command shows that `default-features` is using the flag, yet it's not in the `vcpkg_check_features` call: https://github.com/microsoft/vcpkg/blob/333ba63a16024f05f9172fe403f9eced000389e5/ports/opencv4/portfile.cmake#L63-L96

Can some describe why this line is added, and what it means in context? It appears as if the feature is always "OFF" in opencv4. I'm looking for confirmation that this is the case... But also, why. 

@cenit

**Comment by cenit** [comment]
it's because once there was no "platform" nor "supports" clauses in the manifest. Since the "dnn" feature cannot work in all possible os/arch combinations, it was handled manually like all the others on that line.
In the latest PR in which i am updating opencv i am using all the new features of the manifest to handle all features with vcpkg_check_features

**Reply by gblikas** [reply]
Is the PR up?

**Reply by cenit** [reply]
#38928

**Comment by dg0yt** [comment]
> Can some describe why this line is added, and what it means in context? It appears as if the feature is always "OFF" in opencv4. I'm looking for confirmation that this is the case...

It is not always `OFF`. Just look for `BUILD_opencv_dnn` in `portfile.cmake`. To get started, just look at the lines after initialization:
https://github.com/microsoft/vcpkg/blob/333ba63a16024f05f9172fe403f9eced000389e5/ports/opencv4/portfile.cmake#L98-L105

**Reply by gblikas** [reply]
“dnn” is not in FEATURES

**Reply by dg0yt** [reply]
`dnn` is in the `FEATURES` variable iff you install feature "dnn".
https://github.com/microsoft/vcpkg/blob/9a6da16845eca8d6ed70be416c1acbd206894c7f/ports/opencv4/vcpkg.json#L103-L114

**Reply by gblikas** [reply]
Won't the lack of `"dnn"` in `vcpkg_check_features` cause it not to be detected? Maybe Im confused how that works.

**Reply by dg0yt** [reply]
`vcpkg_check_features` is just a convenience function which uses the top-level `FEATURES` variable as an *input*. The top-level `FEATURES` variable is set by vcpkg tool according to the installation task.

**Reply by gblikas** [reply]
Thank you so much! If only I could read: 
> Check if one or more features are a part of a package installation.
 https://learn.microsoft.com/en-us/vcpkg/maintainers/functions/vcpkg_check_features
 
 lol.

**Reply by dg0yt** [reply]
Well, the `FEATURES` parameter to that function is not the same as the top-level `FEATURES` variable.
(Edited my comments to make this clear.)

**Reply by gblikas** [reply]
> uses the top-level FEATURES variable as an input.

@dg0yt  Ahhh... I was under the impression that `FEATURES` was the `APPEND`'d list, having consumed the `<feature-name> <feature-var>` tuples below it, and was "unique"; really, it just a keyword, correct?

This would be a great thing to elaborate in the documentation, I think. Totally makes sense, but wasn't clear to me as a new user.

**Reply by dg0yt** [reply]
As a user, you shouldn't need to look at the portfile cmake code at all. As a port author, opencv4 might not be a suitable starting point for learning. Note opencv4's history, list of features, and network of dependencies.

`vcpkg_check_features` covers a basic mapping of features to boolean cmake variables. The documentation describes the observable behavior: which variables it sets, subject to `<feature-name>`. (Fine print: It doesn't only set `<out-var>` but also each `<feature-var>`.) Implementation details or alternative processing of `${FEATURES}` are somewhat out of scope for this documentation.

**Reply by gblikas** [reply]
> Note opencv4's history, list of features, and network of dependencies.

yeah... the biggest issue is that multiple platforms default OpenCV's `ocv_option` to different things. So, you cannot use a unified `vcpkg_check_features`... I think that's exactly what @cenit was saying: https://github.com/microsoft/vcpkg/discussions/40455#discussioncomment-10344028
